lib/autoupdate/rollout/strategy_haltonerror.go

```diff
@@ -157,6 +189,138 @@ func (h *haltOnErrorStrategy) progressRollout(ctx context.Context, spec *autoupd
 	return nil
 }
 
+const (
+	canaryCount     = 5
+	canaryThreshold = 20
+)
+
+func shouldUseCanaries(currentCount int) bool {
+	// in the future we might change this logic to be a multiple of the required canary count
+	// and make the canary count dynamic
+	return currentCount >= canaryThreshold
+}
+
+func (h *haltOnErrorStrategy) startGroup(ctx context.Context, group *autoupdate.AutoUpdateAgentRolloutStatusGroup, now time.Time, agentCount int, status *autoupdate.AutoUpdateAgentRolloutStatus) {
+	group.InitialCount = uint64(agentCount)
+
+	if !shouldUseCanaries(agentCount) {
+		h.log.DebugContext(ctx, "Skipping canary rollout, transitioning directly to the active state", "group", group.Name)
+		setGroupState(group, autoupdate.AutoUpdateAgentGroupState_AUTO_UPDATE_AGENT_GROUP_STATE_ACTIVE, updateReasonCanStart, now)
+		return
+	}
+
+	setGroupState(group, autoupdate.AutoUpdateAgentGroupState_AUTO_UPDATE_AGENT_GROUP_STATE_CANARY, updateReasonCanStart, now)
+	// This is a small optimization, as we just transitioned into the canary state we can sample canaries.
+	// This will allow us to start updating without having to wait for the next reconciliation cycle.
+	if err := h.sampleCanaries(ctx, group, status); err != nil {
+		h.log.WarnContext(ctx, "Failed to sample canaries", "group", group.Name)
+	}
+}
+
+func (h *haltOnErrorStrategy) sampleCanaries(ctx context.Context, group *autoupdate.AutoUpdateAgentRolloutStatusGroup, status *autoupdate.AutoUpdateAgentRolloutStatus) error {
+	if group.CanaryCount == 0 {
+		group.CanaryCount = canaryCount
+	}
+	// Check if we need to pick more canaries
+	if len(group.Canaries) < int(group.CanaryCount) {
```

Comment:
Nit: This looks fine but it does give me a bit of worry. I think at first glance, it looks like it's a little liable to a developer mistake in the future leading to an endless loop. I'd almost be tempted to inline `injectCanaries` 😓  But if there's not a clearer/nicer way of doing this, then that's fine.